src/fal/packages/environment.py

```diff
@@ -0,0 +1,157 @@
+from __future__ import annotations
+
+import hashlib
+import json
+import shutil
+import subprocess
+import threading
+from collections import defaultdict
+from contextlib import contextmanager
+from dataclasses import dataclass, field
+from functools import cached_property, partial
+from pathlib import Path
+from typing import (
+    Any,
+    Callable,
+    ClassVar,
+    ContextManager,
+    DefaultDict,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Tuple,
+)
+
+from platformdirs import user_cache_dir
+
+import fal.packages._run_hook as _hook_runner_module
+from faldbt.project import FalDbt
+from fal.planner.tasks import SUCCESS, FAILURE
+
+_BASE_CACHE_DIR = Path(user_cache_dir("fal", "fal"))
+_BASE_CACHE_DIR.mkdir(exist_ok=True)
+
+_BASE_VENV_DIR = _BASE_CACHE_DIR / "venvs"
+_BASE_VENV_DIR.mkdir(exist_ok=True)
+
+# TODO: for dev versions, we need a better solution.
+_IS_DEV_VERSION = True
+
+
+def _get_default_requirements() -> Iterator[Tuple[str, Optional[str]]]:
+    import importlib_metadata
+
+    # TODO: for dev versions, we need a better solution.
+    if _IS_DEV_VERSION:
+        yield "git+https://github.com/fal-ai/fal@batuhan/fea-318-use-the-environment-option-for-local", None
+    else:
+        yield "fal", importlib_metadata.version("fal")
+    yield "dbt-core", importlib_metadata.version("dbt-core")
+
+    for dynamic_dependency in importlib_metadata.distributions():
+        if dynamic_dependency.name.startswith("dbt-"):
+            yield dynamic_dependency.name, dynamic_dependency.version
+
+
+HookRunner = Callable[[FalDbt, Path, str, Dict[str, Any], int], int]
+
+
+@contextmanager
+def _clear_on_fail(path: Path) -> Iterator[None]:
+    try:
+        yield
+    except Exception:
+        shutil.rmtree(path)
+        raise
+
+
+class BaseEnvironment:
+    def setup(self) -> ContextManager[HookRunner]:
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class VirtualPythonEnvironment(BaseEnvironment):
+    requirements: List[str] = field(default_factory=list)
+
+    # TODO: This is not intra-process safe, so if you have 2 running
+    # fal processes that tries to create the same virtual environment
+    # *at the same time* it might fail (the behavior is undefined)·
+    #
+    # We might think about introducing a file-lock, which would
+    # allow this scenerio in the future.
+    _VENV_LOCKS: ClassVar[DefaultDict[threading.Lock]] = defaultdict(threading.Lock)
+
+    def __post_init__(self) -> None:
+        self.requirements.extend(
+            f"{key}=={value}" if value else key
+            for key, value in _get_default_requirements()
```

Comment:
OK, I would say let's ignore fal or dbt versions and log a warning for them

Reply:
The parsing of dependencies would be a bit tricky (`fal==5.0`, `fal!=5.0`, and a lot of other variations) though we can just let pip handle it. It already raises a warning if you have a dependency conflict, and I think it will handle this case too. If it starts creating confusions, we can think about it later I think.
 
```console
 $ pip install pyjokes==0.5.0 pyjokes==0.6.0                                                                                                                                                                                                                                         5h54m
Collecting pyjokes==0.5.0
  Using cached pyjokes-0.5.0-py2.py3-none-any.whl (12 kB)
ERROR: Cannot install pyjokes==0.5.0 and pyjokes==0.6.0 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested pyjokes==0.5.0
    The user requested pyjokes==0.6.0
```

Reply:
Cool, let's do that!